### DISCUSSION TITLE
How to do Co-op through WebGL [OUTDATED]

### DISCUSSION THREAD

**Discussion by DerpedCrusader** [nzp-team/nzportable#400] [Tutorials]
This is all out of date, but it want to keep this up mainly just because it's cute

~~New instructions are in this video
 https://fxtwitter.com/NZPTeam/status/1729603466824216618/video/1~~

THERE'S A SERVER BROWSER NOW! Just host and people can join from there!

# old outdated text below 

A lot of work has gone into making Co-op be a much more enjoyable experience

so, while a complete match maker/browser is still a long way off, I decided to make a little guide to show you how to set a server up for you and your friends!

some clarifying notes, 

-This guide is *ONLY* for the WebGL build of the game
-if you see a bug that's MP exclusive, make sure to report it and label it as such, 
-Have fun

# [CLICK ME FOR THE WEBGL BUILD](https://nzp-team.github.io/latest/game.html)

so, You're all loaded up and in the main menu of the webGL build

press F11 for full screen... at least in most browsers....

# FIRST

We're going to start with what the host needs to do! I'll let you know what the clients (your friends) will have to do when the time comes

we need to open the console up

1: Press the ~ key on your keyboard, it should be to the left of the 1 on English layout keyboards
![image](https://github.com/nzp-team/nzportable/assets/13917665/15099e0b-7695-414d-9cd1-0351005b07cc)

once you do this, a screen should pull down with some text, and a spot for you to type!
![image](https://github.com/nzp-team/nzportable/assets/13917665/749be30c-c2c1-4a82-842d-48128e04f1f5)

# SECOND

Gotta change that name! or else you're going to look like a FOOL!

Type in `name xxxxxx` where the x's are your player name, keep in mind, this is purely cosmetic, and you can put anything

![image](https://github.com/nzp-team/nzportable/assets/13917665/9b61f49a-528a-49fd-909c-3a58499c9261)

and press enter

I know it's kind of trivial, but its important for ease of play

# THIRD

KEEP THAT CONSOLE OPEN!!!

we're going to type in a command, this will make it so other players will be able to connect to your server!

type in `sv_public 1` and press enter

![image](https://github.com/nzp-team/nzportable/assets/13917665/728eba16-36cf-4b1f-b06e-f23c645c9743)

yes, you will have to do this every time per session

# FOURTH

close the console by pressing the same button you did initially `/~

click on solo option in the main menu, and then whichever map you want to play, including the included custom maps!

So, I'm using NDU, the classic

![image](https://github.com/nzp-team/nzportable/assets/13917665/b09d98ca-043a-4e23-bf00-a7a74bfa1ce8)

when you start the game up if you press the console button, the shade should pull down, and there's something super specific you need to look for, and that the numbers after the final `/`

![image](https://github.com/nzp-team/nzportable/assets/13917665/6b8a2cb4-f671-4ece-bf4c-bd76ecb3b0b6)

my number is 575, but yours will be different, it might even be 4 digits+ as more and more people play!

That number, is the key to getting your friends to connect to your server, so make sure to write it down / message it to your friends

now, let's get them in the game!

# FIFTH 

Have your friends (the clients) open the webGL build of the game, just copy/paste the link for them

first, have them open the console, and do `name xxxxx` command for themselves, once again, only cosmetic, it can be anything you want, preferably something different from your friend's

and here's where the magic happens

have them type `connect /XXX` with the X's being the number earlier, in my case, its 575

![image](https://github.com/nzp-team/nzportable/assets/13917665/0f32886a-292c-4f56-8484-62962ea33eb8)

if everything went well, you should be able to see each other!

![image](https://github.com/nzp-team/nzportable/assets/13917665/ac4404b3-429f-4aed-95dd-5a63ce8bee23)
(Split screen not supported, this image is for example purposes)

Now, you might have some issues, so here's some common ones!

Q: Me (the host) is fine, but my friend is floating around or spectating?
A: that means they joined too late, it happens often, just finish the current round, and they should spawn in at the start of the next round,

this can be mitigated if the host pauses (esc) right when the match starts... ~~but that causes a few other issues, which are listed below~~

~~Q: We're all stuck inside each other!
A: not all maps at the moment have proper multiplayer spawns, there is a fix in the works, but it's not out yet, so here's what you can do:~~

~~make sure everyone that is connecting is staggering when they connect, and you can pause after each person moves~~

Should be fixed!

Q: when i un paused it didn't un pause?
A: simply open the console (~) and type `pause` hit enter, and it should work just fine

Q: everything is laggy
A: this can happen for a variety of reasons, like using Wifi, or long distance from the players, NZP doesn't use any form of player prediction, so it can be especially brutal at times

**Comment by zqpvr** [comment]
I am stuck at the part where you say "click on solo -> and whichever map you want to play, including the included custom maps!" can you please explain more detail or add pictures.

**Reply by DerpedCrusader** [reply]
You can just start any map, it doesn't get more simple than that, I can put some images up, but it's pretty self explanatory

**Reply by zqpvr** [reply]
it seems to actually startup the thing but now it doesn't let friends connect it just keeps looping on connecting /XXX
